packages/jupyter-ai/jupyter_ai/personas/persona_manager.py

```diff
@@ -0,0 +1,174 @@
+        to the constructor.
+        """
+        if hasattr(self, "_personas") or len(PersonaManager._persona_classes) == 0:
+            return
+        persona_classes = PersonaManager._persona_classes
+        assert isinstance(persona_classes, list)
+
+        self.log.info(
+            f"PENDING: Initializing AI personas for chat room '{self.ychat.get_id()}'..."
+        )
+        start_time_ns = time_ns()
+
+        personas: dict[str, BasePersona] = {}
+        for Persona in persona_classes:
+            persona = Persona(
+                ychat=self.ychat,
+                manager=self,
+                config=self.config_manager,
+                log=self.log,
+            )
+            if persona.id in personas:
+                class_name = persona.__class__.__name__
+                self.log.warning(
+                    f"  - WARNING: Skipping persona '{persona.name}' from '{persona.__module__}' because another persona has an identical ID '{persona.id}'. "
+                    + f"Personas must all have unique IDs. Please rename the persona class from '{class_name}' to something unique to dismiss this warning."
+                )
+                continue
+            else:
+                self.log.info(
+                    f"  - Initialized persona '{persona.name}' (ID: '{persona.id}')."
+                )
+                personas[persona.id] = persona
+
+        elapsed_time_ms = (time_ns() - start_time_ns) // 1000
+        self.log.info(
+            f"SUCCESS: Initialized {len(personas)} AI personas for chat room '{self.ychat.get_id()}'. Time elapsed: {elapsed_time_ms}ms."
+        )
+        return personas
+
+    @property
+    def personas(self) -> dict[str, BasePersona]:
+        return self._personas
+
+    def get_mentioned_personas(self, new_message: Message) -> list[BasePersona]:
+        mentioned_ids = set(new_message.mentions)
+        persona_list: list[BasePersona] = []
+
+        for mentioned_id in mentioned_ids:
+            if mentioned_id in self.personas:
+                persona_list.append(self.personas[mentioned_id])
+
+        return persona_list
+
+    def route_message(self, new_message: Message):
+        """
+        Method that routes an incoming message to the correct persona by calling
+        its `process_message()` method.
+
+        - (TODO) If the chat contains only one persona & one user, then this
+        method routes all new messages to that persona.
+
+        - Otherwise, this method only routes new messages to personas that are
+        `@`-mentioned in the message.
+        """
+        mentioned_personas = self.get_mentioned_personas(new_message)
+        if not len(mentioned_personas):
+            return
+
+        mentioned_persona_names = [persona.name for persona in mentioned_personas]
+        self.log.info(
+            f"Received new user message mentioning the following personas: {mentioned_persona_names}."
+        )
```

Comment:
optional: isn't this a lot of stdout, maybe `debug` ?

Reply:
I think it's good to be verbose with our logs this early in v3 development. We can address clean logging in the future, probably in the RC phase.

Reply:
Opened a new issue for you: #1339.